src/main/scala/com/fulcrumgenomics/bam/IdentifyPrimers.scala

```diff
@@ -0,0 +1,1027 @@
+    }
+    // The number of templates to batch across threads
+    val majorBatchSize  = math.min(batchSize * threads * 4, maxTemplatesInRam.getOrElse(Int.MaxValue))
+    val readingProgress = ProgressLogger(this.logger, verb="read", unit=batchSize*threads*2)
+
+    // Batch templates to process them in individual threads.
+    val outputIterator: Iterator[Seq[Template]] = if (threads > 1) {
+      logger.info(f"Batching $majorBatchSize%,d templates with $batchSize%,d templates per thread.")
+
+      import com.fulcrumgenomics.commons.CommonsDef.ParSupport
+      // Developer Note: Iterator does not support parallel operations, so we need to group together records into a
+      // [[List]] or [[Seq]].  A fixed number of records are grouped to reduce memory overhead.
+      iterator
+        .grouped(majorBatchSize)
+        .flatMap { templates =>
+          templates
+            .grouped(batchSize)
+            .toStream
+            .parWith(threads, fifo = false)
+            .map { templates => processBatch(templates, metricCounter, readingProgress) }
+            .seq // ensures that the only parallelism is processBatch
+            .toIterator
+        }
+    }
+    else {
+      logger.info(f"Batching $batchSize%,d templates.")
+      iterator
+        .grouped(batchSize)
+        .map { templates => processBatch(templates, metricCounter, readingProgress) }
+    }
+
+    // Write the results
+    val writingProgress = ProgressLogger(this.logger, "written", unit=1000000)
+    outputIterator
+      .flatMap(_.flatMap(_.allReads))
+      .foreach { rec =>
+        writingProgress.record(rec)
+        out += rec
+      }
+
+    val rate = numAlignments.get() / readingProgress.getElapsedSeconds.toDouble
+    logger.info(f"Performed ${numAlignments.get()}%,df full alignments in total ($rate%,.2f alignments/second).")
+    logger.info(f"Wrote ${readingProgress.getCount}%,d records.")
+
+    in.safelyClose()
+    out.close()
+
+
+    // Write metrics
+    // Detailed metrics
+    {
+      val total   = metricCounter.total.toDouble
+      val metrics = metricCounter.map { case (metric, count) =>
+        metric.copy(count = count, frac = count / total)
+      }.toSeq.sortBy(-_.count)
+      Metric.write(PathUtil.pathTo(this.metrics + ".detailed.txt"), metrics)
+    }
+    // Summary metrics
+    Metric.write(PathUtil.pathTo(this.metrics + ".summary.txt"), IdentifyPrimersMetric(metricCounter))
+  }
+
+  /** Creates a new [[InteractiveAlignmentScorer]]. */
+  private def newAligner: InteractiveAlignmentScorer[Primer, SamRecordAlignable] = {
+    InteractiveAlignmentScorer(matchScore, mismatchScore, gapOpen, gapExtend, AlignmentMode.Glocal, this.kswExecutable)
+  }
+
+  // NB: batching alignment inputs to the aligner (i.e. ksw) is empirically faster than given them all or just one at a time.
+  private def completePrimerMatch(matchOption: Option[Either[PrimerMatch, Seq[AlignmentTask[Primer, SamRecordAlignable]]]],
```

Comment:
Wouldn't this be clearer as two methods rather that one which takes this crazy `Option[Either[PrimerMatch, Seq[AlignmentTask[Primer, SamRecordAlignable]]]]`?

Reply:
Wait .. so it doesn't even do anything if you pass in a `Some(Left(PrimerMatch))`?  Why is that even here then?